### ISSUE TITLE
Implement the sample-based NeuRD

### ISSUE THREAD

**Issue by godmoves** [deepmind/open_spiel#891]
Hi, I have just read the [DeepNash](https://arxiv.org/abs/2206.15378) paper and I am quite interested in the NeuRD algorithm. But it seems the sample-based version is not implemented in the OpenSpiel repository, so I am trying to do it myself.

I find there are some discussions about it in #148 and #316, so I just follow the instructions in #316 and add a **PyTorch** version. I choose to use the **regular policy entropy** instead of the q-value entropy used in https://arxiv.org/abs/2002.08456 because I still have some questions about it and decide to leave it for a future task.

To compare the results between different algorithms, I run each algorithm 5 times using the OpenSpiel PyTorch code and reproduce the results in [Figure 3](https://arxiv.org/abs/1908.09453) (except NFSP). The hyperparameters are from #155. The final performances appear to be basically the same as in the paper, except for the RPG, which is **worse**.

Due to the different entropy types used, I need to do more hyperparameter searches before I finally confirm the performance of my implementation of NeuRD. I will update the results once I have done that.

![image](https://user-images.githubusercontent.com/25997471/180639084-c026e79b-2c70-43e7-8c26-057db0648209.png)

**Comment by godmoves** [comment]
To verify the RPG result of the PyTorch code, I also ran an extra experiment on the Tensorflow code. But the result is still worse than the value reported in the paper.

It is mentioned that the learning rate is decayed to 0 in the [RPG paper](https://arxiv.org/abs/1810.09026). Is this also used in the OpenSpiel experiments?

> These updates were done using separate SGD optimizers with their respective learning rates of fixed 0.001 for policy evaluation, and annealed from a starting learning rate to 0 over 20M steps for policy improvement.

There is a temperature for softmax, too.

> In addition, the agent’s policy is controlled by a(n inverse) temperature added as part of the softmax operator. The temperature is annealed from 1 to 0 over 1M steps to ensure adequate state space coverage.

By the way, the hyperparameter I used, and all learning rate is fixed during the training:
```
loss_str:rpg
batch_size:64
critic_lr:0.01
pi_lr:0.01
entropy_cost:0.01
num_critic_before_pi:32
```
![image](https://user-images.githubusercontent.com/25997471/180682456-c091de4b-637d-4f2a-bd98-9da4959d954d.png)

**Comment by lanctot** [comment]
Hi @godmoves, this is awesome! It's been repeatedly asked for (as you found out), so it would be great to finally include in the library.

For the RPG results, I'd recommend looking at this paper: https://arxiv.org/abs/2103.00187. I think they found a different set of hyper-parameters worked better than the defaults we have (which I think may have been tuned on Kuhn poker) and the ones from the original paper because they used a slightly different implementation of the domains. The values in that paper should be reliable because they used the same OpenSpiel implementations which have mainly remain unchanged.

Can you submit a pull request so we can look at the code?

I'd like to loop in the original authors regarding your questions about NeuRD itself, and that might be easier if we see the code. What do you think?

**Comment by godmoves** [comment]
@lanctot Thank you for your reply. I submit a pull request in #892.

For the RPG, I ran a new experiment with policy learning rate decay from the initial value to 0, and its performance is better than fixed learning rate. I will use the hyperparameter in https://arxiv.org/abs/2103.00187 to run an extra RPG check.

![image](https://user-images.githubusercontent.com/25997471/180918027-00f5ef7b-f4ec-4e0a-ac5e-2fb5946606ee.png)

**Comment by godmoves** [comment]
Unfortunately, I can not reproduce the RPG result on Kuhn poker in https://arxiv.org/abs/2103.00187 even using their [open-sourced code](https://github.com/aicenter/openspiel_reproductions). I don't know if there are some changes in the OpenSpiel codebase that affect the behavior of RPG.

But the good news is that NeuRD's performance is better than RMPG/RPG/QPG on Kuhn Poker both in speed and exploitability. 

The neural network used is a 2-layer MLP with 128 units in each layer. The hyperparameters used for each algorithm are listed below and most of the values are the same as in #155.

|name|batch_size|critic_lr|entropy_cost|num_critic_before_pi|pi_lr|
|------|----------|-------|-------------|---------------------|----|
|RPG|64|0.01|0.01|32|0.01|
|QPG|64|0.01|0.01|8|0.01|
|RMPG|64|0.01|0.01|8|0.01|
|NeuRD|32|0.01|0.01|32|0.005|
|RPG 2103.00187|4|0.01|0.1|128|0.01|

I ran each algorithm 5 times and 5e6 steps. Here are the results.
![image](https://user-images.githubusercontent.com/25997471/181192174-e47d8590-f367-4b05-b4e8-093b329936df.png)

**Comment by lanctot** [comment]
Wow, reproducibility is hard :'(

There have been several Tensorflow (and other python package) upgrades since then. Could be something numerical? I can't think of any changes to the policy gradient code that would have caused it. 

Either way, it is good to know about. I will keep it in mind and follow-up if I think of anything. Meanwhile, thanks again for contributing the sampled-based NeuRD loss.

**Comment by lanctot** [comment]
Closing as resolved from: https://github.com/deepmind/open_spiel/pull/892

Thanks for the contribution!
